Thanks for the review!

Safer C++ check failure seems infra-related:

```
/Volumes/Data/worker/macOS-Safer-CPP-Checks-EWS/build/WebKitBuild/Release/usr/local/include/wtf/NeverDestroyed.h:155:40: warning: Unsafe cast from type 'type' to an unrelated type 'QualifiedName' [alpha.webkit.MemoryUnsafeCastChecker]
  155 |         return const_cast<PointerType>(reinterpret_cast<const T*>(&m_storage));
      |                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
fatal error: error in backend: IO failure on output stream: No space left on device
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
```

--- a/LayoutTests/editing/selection/ios/bidi-visually-contiguous-selection-5-expected.txt
+++ b/LayoutTests/editing/selection/ios/bidi-visually-contiguous-selection-5-expected.txt
@@ -1,0 +1,12 @@
+Verifies that the selection does not unnecessarily expand to the entire paragraph when selecting bidi text surrounded by bidi embedding characters.
+
+On success, you will see a series of "PASS" messages, followed by "TEST COMPLETE".
+
+
+PASS selectionRects.length is 1
+PASS selectionRects[0].width is <= bounds.width - 1
+PASS getSelection().toString() is "iPhone"
+PASS successfullyParsed is true
+
+TEST COMPLETE
+

--- a/LayoutTests/editing/selection/ios/bidi-visually-contiguous-selection-5.html
+++ b/LayoutTests/editing/selection/ios/bidi-visually-contiguous-selection-5.html
@@ -1,0 +1,38 @@
+<!DOCTYPE html> <!-- webkit-test-runner [ useFlexibleViewport=true VisuallyContiguousBidiTextSelectionEnabled=true ] -->
+<html>
+<head>
+<script src="../../../resources/js-test.js"></script>
+<script src="../../../resources/ui-helper.js"></script>
+<meta name="viewport" content="width=device-width, initial-scale=1">
+<meta charset="utf-8">
+<style>
+body, html {
+    font-size: 20px;
+    font-family: system-ui;
+}
+</style>
+<script>
+jsTestIsAsync = true;
+
+addEventListener("load", async () => {
+    description("Verifies that the selection does not unnecessarily expand to the entire paragraph when selecting bidi text surrounded by bidi embedding characters.");
+
+    bounds = document.getElementById("target").getBoundingClientRect();
+
+    await UIHelper.longPressAtPoint(bounds.left + 20, bounds.top + (bounds.height / 2));
+    selectionRects = await UIHelper.waitForSelectionToAppear();
+    await UIHelper.ensurePresentationUpdate();
+
+    shouldBe("selectionRects.length", "1");
+    shouldBeLessThanOrEqual("selectionRects[0].width", "bounds.width - 1");
+    shouldBeEqualToString("getSelection().toString()", "iPhone");
+
+    document.querySelector("p[dir='rtl']").remove();
+    finishJSTest();
+});
+</script>
+</head>
+<body>
+    <p dir="rtl"><span id="target">&#x202b;أُرسلت من الـ iPhone&#x202c;</span></p>
+</body>
+</html>

--- a/Source/WebCore/editing/Editing.cpp
+++ b/Source/WebCore/editing/Editing.cpp
@@ -1318,19 +1318,33 @@ static void forEachRenderedBoxBetween(const RenderedPosition& first, const Rende
         return;
     }
 
-    bool foundEndpoint = false;
+    bool foundOneEndpoint = false;
     for (auto box = first.lineBox()->lineLeftmostLeafBox(); box; box = box->nextLineRightwardOnLine()) {
-        bool atEndpoint = box == first.box() || box == second.box();
-        if (!atEndpoint && !foundEndpoint)
+        bool atFirstEndpoint = box == first.box();
+        bool atSecondEndpoint = box == second.box();
+        bool atEndpoint = atFirstEndpoint || atSecondEndpoint;
+        bool atLastEndpoint = atEndpoint && foundOneEndpoint;
+        if (!atEndpoint && !foundOneEndpoint)
+            continue;
+
+        foundOneEndpoint = true;
+
+        bool shouldSkipBox = [&] {
+            if (!atEndpoint)
+                return false;
+
+            auto& position = (atFirstEndpoint ? first : second);
+            return atLastEndpoint ? position.atLeftmostOffsetInBox() : position.atRightmostOffsetInBox();
+        }();
+
+        if (shouldSkipBox)
             continue;
 
         if (callback(box) == IterationStatus::Done)
             return;
 
-        if (atEndpoint && foundEndpoint)
+        if (atLastEndpoint)
             return;
-
-        foundEndpoint = true;
     }
 }
 
@@ -1422,7 +1436,7 @@ static std::optional<SimpleRange> makeVisuallyContiguousIfNeeded(const SimpleRan
         }
     }
 
-    if (adjustedEndpoints)
+    if (adjustedEndpoints && !adjustedRange.collapsed())
         return adjustedRange;
 
     return std::nullopt;

--- a/Source/WebCore/editing/RenderedPosition.h
+++ b/Source/WebCore/editing/RenderedPosition.h
@@ -72,6 +72,9 @@ public:
     Position positionAtLeftBoundaryOfBiDiRun() const;
     Position positionAtRightBoundaryOfBiDiRun() const;
 
+    bool atLeftmostOffsetInBox() const { return m_box && m_offset == m_box->leftmostCaretOffset(); }
+    bool atRightmostOffsetInBox() const { return m_box && m_offset == m_box->rightmostCaretOffset(); }
+
     IntRect absoluteRect(CaretRectMode = CaretRectMode::Normal) const;
 
     std::optional<BoundaryPoint> boundaryPoint() const;
@@ -82,8 +85,6 @@ private:
 
     InlineIterator::LeafBoxIterator previousLeafOnLine() const;
     InlineIterator::LeafBoxIterator nextLeafOnLine() const;
-    bool atLeftmostOffsetInBox() const { return m_box && m_offset == m_box->leftmostCaretOffset(); }
-    bool atRightmostOffsetInBox() const { return m_box && m_offset == m_box->rightmostCaretOffset(); }
     bool atLeftBoundaryOfBidiRun(ShouldMatchBidiLevel, unsigned char bidiLevelOfRun) const;
     bool atRightBoundaryOfBidiRun(ShouldMatchBidiLevel, unsigned char bidiLevelOfRun) const;
 

